### PR TITLE
Azure - ARM Read/Write rate limiting

### PR DESCRIPTION
**Pull request by khenidak**
**What this PR does / why we need it**:

Azure cloud provider currently runs with:
1. Single ARM rate limiter for both `read [put/post/delete]` and `write` operations, while ARM provide [different rates for read/write] (https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-manager-request-limits). This causes write operation to stop even if there is available write request quotas. 
2. Cloud provider uses rate limiter's `Accept()` instead of `TryAccept()` This causes control loop to wait for prolonged tike `in case of no request quota available` for **all** requests even for those does not require ARM interaction. A case for that the `Service` control loop will wait for a prolonged time trying to create `LoadBalancer` service even though it can fail and work on the next service which is `ClusterIP`. This PR moves cloud provider tp `TryAccept()`

**Which issue(s) this PR fixes**:
Fixes # https://github.com/kubernetes/kubernetes/issues/58770

**Special notes for your reviewer**:
`n/a`

**Release note**:

```release-note
- Separate current ARM rate limiter into read/write
- Improve control over how ARM rate limiter is used within Azure cloud provider
```

cc @jackfrancis (need your help carefully reviewing this one) @brendanburns @jdumars

### CODE REVIEW

**Comment by khenidak** [comment]
@jackfrancis @brendanburns This is ready for review now.

File: pkg/cloudprovider/providers/azure/azure_client.go [hunk 1]
```diff
@@ -168,7 +192,11 @@ func (az *azVirtualMachinesClient) CreateOrUpdate(resourceGroupName string, VMNa
 }
 
 func (az *azVirtualMachinesClient) Get(resourceGroupName string, VMName string, expand compute.InstanceViewTypes) (result compute.VirtualMachine, err error) {
-	az.rateLimiter.Accept()
+	err = createArmRateLimitErr(false, "VMGet")
+	if !az.rateLimiterReader.TryAccept() {
+		return
```
Comment (jackfrancis) [inline pkg/cloudprovider/providers/azure/azure_client.go]: Educate me in golang. :)

By implicitly returning (simple `return` without params) both named return types, what are we returning as the value of `result` in the error case?

File: pkg/cloudprovider/providers/azure/azure_client.go [hunk 2]
```diff
@@ -181,7 +209,11 @@ func (az *azVirtualMachinesClient) Get(resourceGroupName string, VMName string,
 }
 
 func (az *azVirtualMachinesClient) List(resourceGroupName string) (result compute.VirtualMachineListResult, err error) {
-	az.rateLimiter.Accept()
+	err = createArmRateLimitErr(false, "VMList")
```
Comment (jackfrancis) [inline pkg/cloudprovider/providers/azure/azure_client.go]: Could we optimize by putting these `err` constructor invocations inside the failure condition if block (as they are only used if the failure condition is met)?

**Comment by jackfrancis** [comment]
@khenidak Added a couple comments for thought. Overall lgtm, audited all the rate limit injections (esp. the write ones which require us to compose a zero value return value) and they look correct/sane.

**Comment by khenidak** [comment]
@jackfrancis updated with your code comments PTAL
@feiskyer I have had to merge the VMSS code. can you do a sanity check?

**Review by jackfrancis — APPROVED** [review]
lgtm

**Comment by jdumars** [comment]
@feiskyer if you're good with this implementation can you add the LGTM so we can get this merged?

File: pkg/cloudprovider/providers/azure/azure_client.go [hunk 3]
```diff
@@ -33,6 +34,15 @@ import (
 	"k8s.io/client-go/util/flowcontrol"
 )
 
+// Creates an error for rate limiting errors
+func createArmRateLimitErr(isWrite bool, opName string) error {
```
Comment (brendandburns) [inline pkg/cloudprovider/providers/azure/azure_client.go]: Given that the purpose of this is to return this error in a channel, perhaps having this signature being:

Also nit, golang style says Arm should be ARM.

`func createARMRateLimitError(isWrite bool, opName string) (err, chan error) {`

Comment (brendandburns) [reply pkg/cloudprovider/providers/azure/azure_client.go]: ah, I actually see that you use this elsewhere, perhaps two functions:

```
func createARMRateLimitError(...) error {
```

and 

```
func createARMRateLimitErrChannel() chan error {
```

**Comment by khenidak** [comment]
@brendanburns code comments addressed. Thanks

**Review by feiskyer — APPROVED** [review]
/lgtm

**Pull request merged**